apps/etterlatte-behandling/src/main/kotlin/behandling/domain/Behandling.kt

```diff
@@ -0,0 +1,153 @@
+    abstract val persongalleri: Persongalleri
+    abstract val kommerBarnetTilgode: KommerBarnetTilgode?
+    abstract val vilkaarUtfall: VilkaarsvurderingUtfall?
+    abstract val virkningstidspunkt: Virkningstidspunkt?
+
+    protected val logger: Logger = LoggerFactory.getLogger(this.javaClass.name)
+    val oppgaveStatus get() = OppgaveStatus.from(status)
+
+    private val kanRedigeres: Boolean
+        get() = this.status.kanRedigeres()
+
+    protected fun <T : Behandling> hvisRedigerbar(block: () -> T): T {
+        if (kanRedigeres) {
+            return block()
+        } else {
+            logger.info("behandling ($id) med status $status kan ikke redigeres")
+            throw TilstandException.UgyldigtTilstand
+        }
+    }
+
+    protected fun <T : Behandling> hvisTilstandEr(behandlingStatus: BehandlingStatus, block: () -> T): T {
+        if (status == behandlingStatus) {
+            return block()
+        } else {
+            logger.info("Ugyldig operasjon på behandling ($id) med status $status")
+            throw TilstandException.UgyldigtTilstand
+        }
+    }
+
+    protected fun <T : Behandling> hvisTilstandEr(behandlingStatuser: List<BehandlingStatus>, block: () -> T): T {
+        if (status in behandlingStatuser) {
+            return block()
+        } else {
+            logger.info("Ugyldig operasjon på behandling ($id) med status $status")
+            throw TilstandException.UgyldigtTilstand
+        }
+    }
+
+    open fun tilOpprettet(): Behandling {
+        throw BehandlingStoetterIkkeStatusEndringException(OPPRETTET)
+    }
+
+    open fun tilVilkaarsvurdert(utfall: VilkaarsvurderingUtfall?): Behandling {
+        throw BehandlingStoetterIkkeStatusEndringException(VILKAARSVURDERT)
+    }
+
+    open fun tilBeregnet(): Behandling {
+        throw BehandlingStoetterIkkeStatusEndringException(BEREGNET)
+    }
+
+    open fun tilFattetVedtak(): Behandling {
+        throw BehandlingStoetterIkkeStatusEndringException(FATTET_VEDTAK)
+    }
+
+    open fun tilAttestert(): Behandling {
+        throw BehandlingStoetterIkkeStatusEndringException(ATTESTERT)
+    }
+
+    open fun tilReturnert(): Behandling {
+        throw BehandlingStoetterIkkeStatusEndringException(RETURNERT)
+    }
+
+    open fun tilIverksatt(): Behandling {
+        throw BehandlingStoetterIkkeStatusEndringException(OPPRETTET)
+    }
+
+    private class BehandlingStoetterIkkeStatusEndringException constructor(
+        behandlingStatus: BehandlingStatus,
+        message: String = "Behandlingen støtter ikke statusendringen til status $behandlingStatus"
+    ) : Exception(message)
+}
+
+internal fun Behandling.toDetaljertBehandling(): DetaljertBehandling {
+    val (soeknadMottatDato, gyldighetsproeving) = when (this) {
+        is Foerstegangsbehandling -> this.soeknadMottattDato to this.gyldighetsproeving
+        // TODO øh 24.10.2022 er det riktig at søknadMottatDato er behandlingOpprettet der vi ikke har søknad?
```

Comment:
Det er vel bare et definisjonsspørsmål? De virker like spør du meg, men jeg ser at det er lagt opp til at det skal være en liten diff når man mottar en søknad:
```
            logger.info("Oppretter en behandling på $sak")
            return Foerstegangsbehandling(
                id = UUID.randomUUID(),
                sak = sak,
                behandlingOpprettet = LocalDateTime.now(),
                sistEndret = LocalDateTime.now(),
                status = BehandlingStatus.OPPRETTET,
                soeknadMottattDato = LocalDateTime.parse(mottattDato),
```

Reply:
Jeg tenkte opprinnelig på om man burde sette dette som `null` der man ikke har en søknad, men det knakk en del antagelser for en stund tilbake om at alle behandlinger har en søknad mottatt dato. Jeg mener egentlig at det burde være null, men det er ikke en veldig sak :p